src/wayland/tablet_manager/tablet_tool.rs

```diff
@@ -0,0 +1,477 @@
+                                } else {
+                                    cb(&desc, CursorImageStatus::Hidden);
+                                };
+                            }
+                        }
+                    }
+                    Request::Destroy => {
+                        // Handled by our destructor
+                    }
+                    _ => {}
+                }
+            });
+
+            let inner = self.inner.clone();
+            wl_tool.assign_destructor(Filter::new(move |instance: ZwpTabletToolV2, _, _| {
+                inner
+                    .borrow_mut()
+                    .instances
+                    .retain(|i| !i.as_ref().equals(&instance.as_ref()));
+            }));
+
+            seat.tool_added(&wl_tool);
+
+            wl_tool._type(tool.tool_type.into());
+
+            let high: u32 = (tool.hardware_serial >> 16) as u32;
+            let low: u32 = tool.hardware_serial as u32;
+
+            wl_tool.hardware_serial(high, low);
+
+            let high: u32 = (tool.hardware_id_wacom >> 16) as u32;
+            let low: u32 = tool.hardware_id_wacom as u32;
+            wl_tool.hardware_id_wacom(high, low);
+
+            if tool.capabilitys.pressure {
+                wl_tool.capability(zwp_tablet_tool_v2::Capability::Pressure);
+            }
+
+            if tool.capabilitys.distance {
+                wl_tool.capability(zwp_tablet_tool_v2::Capability::Distance);
+            }
+
+            if tool.capabilitys.tilt {
+                wl_tool.capability(zwp_tablet_tool_v2::Capability::Tilt);
+            }
+
+            if tool.capabilitys.slider {
+                wl_tool.capability(zwp_tablet_tool_v2::Capability::Slider);
+            }
+
+            if tool.capabilitys.rotation {
+                wl_tool.capability(zwp_tablet_tool_v2::Capability::Rotation);
+            }
+
+            if tool.capabilitys.wheel {
+                wl_tool.capability(zwp_tablet_tool_v2::Capability::Wheel);
+            }
+
+            wl_tool.done();
+            self.inner.borrow_mut().instances.push(wl_tool.deref().clone());
+        }
+    }
+
+    /// Notifify that this tool is focused on a certain surface.
+    ///
+    /// You provide the location of the tool, in the form of:
+    ///
+    /// - The coordinates of the tool in the global compositor space
+    /// - The surface on top of which the tool is, and the coordinates of its
+    ///   origin in the global compositor space.
+    pub fn proximity_in(
+        &self,
+        pos: (f64, f64),
+        focus: (WlSurface, (f64, f64)),
+        tablet: &TabletHandle,
+        serial: Serial,
+        time: u32,
+    ) {
+        self.inner
+            .borrow_mut()
+            .proximity_in(pos, focus, tablet, serial, time)
+    }
+
+    /// Notifify that this tool has left proximity.
+    pub fn proximity_out(&self, time: u32) {
+        self.inner.borrow_mut().proximity_out(time);
+    }
```

Comment:
If `motion()` already handles the `proximity_in/proximity_out` logic, is it necessary to expose these methods as well?

Reply:
It is exposed because tablet tool pointers in contrary to mouse pointers can appear out of tin air, for example if the user lifts a pen above the tablet and moves it to another window (without tablet's knowledge) and then lowers down a pen (triggering tool's proximity_in event) then the pointer will just teleport to the new location. Or even better example would be if a user brings a completely new tool into proximity, then it's not really a `motion` per se, it just appeared out of nowhere.

Although motion event can easily emulate this behavior, I just think that the event handler code is a lot easier to understand if it uses those function directly.
[code that I'm talking about](https://github.com/Smithay/smithay/blob/921ef4339769a7456a7f8bf1e902bdc1b2d3cefb/anvil/src/input_handler.rs#L314)

Now, handling of proximity event in anvil looks like this:
```rs
 match evt.state() {
    ProximityState::In => tool.proximity_in(
        self.pointer_location,
        under,
        &tablet,
        SCOUNTER.next_serial(),
        evt.time(),
    ),
    ProximityState::Out => tool.proximity_out(evt.time()),
}
```
It's really easy to understand at first glance.

Here is how it would look like if `proximity_in` and `proximity_out` were not exposed:
```rs
match evt.state() {
    ProximityState::In => tool.motion(
        self.pointer_location,
        under,
        &tablet,
        SCOUNTER.next_serial(),
        evt.time(),
    ),
    ProximityState::Out => tool.motion(
        (0.0,0.0), // Accually tool outside of proximiti does not have pos, but we have to send something
        None, // Tool out of proximity wil never have anything under it
        &tablet,
        SCOUNTER.next_serial(),
        evt.time(),
    ),
}
```

Or maybe I'm just overthink it, and it's not a such big deal, I'll do whatever you prefer.

Reply:
I see, thanks for the explanation, that makes total sense. :+1: